From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Branson <andrew.branson@jolla.com>
Date: Wed, 5 Aug 2026 18:16:38 +0200
Subject: [sailfishos][sqlite] Use the platform SQLite library

Embedding Gecko into a Qt process with both libmozsqlite3 and the platform
SQLite library makes their unprefixed symbols vulnerable to ELF
interposition.

Link Gecko and its Rust consumers to the platform library, while keeping
carray and sqlite-vec as per-connection static extensions. Avoid changing or
shutting down process-global SQLite state that Qt may already own.

Require ESR140's minimum SQLite version and the platform features Gecko relies
on. Treat core FTS5 as already loaded and omit libmozsqlite3 from packaging.

Generate the JavaScript system-SQLite AppConstant directly from the
configure define.
---
 .../urlbar/private/SuggestBackendRust.sys.mjs |  6 ++-
 browser/installer/package-manifest.in         |  2 +
 build/moz.configure/sqlite.configure          | 18 +++++++++
 config/external/sqlite/moz.build              | 23 +++++++-----
 storage/SQLiteMutex.h                         |  6 +--
 storage/moz.build                             | 20 ++++++++--
 storage/mozStorageConnection.cpp              | 37 +++++++++++++++++--
 storage/mozStorageService.cpp                 | 19 +++++++++-
 storage/test/unit/test_storage_ext_fts5.js    | 36 ++++++++++--------
 third_party/sqlite3/ext/moz.build             | 19 +++++++---
 toolkit/library/moz.build                     |  3 ++
 toolkit/modules/AppConstants.sys.mjs          | 11 +++++-
 toolkit/moz.configure                         |  4 ++
 toolkit/xre/AutoSQLiteLifetime.cpp            | 18 ++++++---
 toolkit/xre/moz.build                         |  2 +
 tools/@types/subs/AppConstants.sys.d.mts      |  2 +
 16 files changed, 176 insertions(+), 50 deletions(-)
 create mode 100644 build/moz.configure/sqlite.configure

diff --git a/browser/components/urlbar/private/SuggestBackendRust.sys.mjs b/browser/components/urlbar/private/SuggestBackendRust.sys.mjs
index 04b2d57b4cf5..dccd55d14d80 100644
--- a/browser/components/urlbar/private/SuggestBackendRust.sys.mjs
+++ b/browser/components/urlbar/private/SuggestBackendRust.sys.mjs
@@ -623,11 +623,13 @@ export class SuggestBackendRust extends SuggestBackend {
     try {
       builder = lazy.SuggestStoreBuilder.init()
         .dataPath(this.#storeDataPath)
-        .remoteSettingsService(this.#remoteSettingsService)
-        .loadExtension(
+        .remoteSettingsService(this.#remoteSettingsService);
+      if (!AppConstants.MOZ_SYSTEM_SQLITE) {
+        builder = builder.loadExtension(
           AppConstants.SQLITE_LIBRARY_FILENAME,
           "sqlite3_fts5_init"
         );
+      }
     } catch (error) {
       this.logger.error("Error creating SuggestStoreBuilder", error);
       return null;
diff --git a/browser/installer/package-manifest.in b/browser/installer/package-manifest.in
index 203811a0fb98..047307a2151d 100644
--- a/browser/installer/package-manifest.in
+++ b/browser/installer/package-manifest.in
@@ -178,9 +178,11 @@
 #endif
 #endif
 @RESPATH@/platform.ini
+#ifndef MOZ_SYSTEM_SQLITE
 #ifndef MOZ_FOLD_LIBS
 @BINPATH@/@DLL_PREFIX@mozsqlite3@DLL_SUFFIX@
 #endif
+#endif
 @BINPATH@/@DLL_PREFIX@lgpllibs@DLL_SUFFIX@
 @BINPATH@/@DLL_PREFIX@gkcodecs@DLL_SUFFIX@
 @BINPATH@/@DLL_PREFIX@mozavutil@DLL_SUFFIX@
diff --git a/build/moz.configure/sqlite.configure b/build/moz.configure/sqlite.configure
new file mode 100644
index 000000000000..2363b7fee66b
--- /dev/null
+++ b/build/moz.configure/sqlite.configure
@@ -0,0 +1,18 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+system_lib_option(
+    "--with-system-sqlite",
+    help="Use system SQLite (located with pkgconfig)",
+    when=use_pkg_config,
+)
+
+system_sqlite = pkg_check_modules(
+    "SQLITE", "sqlite3 >= 3.53.2", when="--with-system-sqlite"
+)
+
+set_config("MOZ_SYSTEM_SQLITE", True, when=system_sqlite)
+set_define("MOZ_SYSTEM_SQLITE", True, when=system_sqlite)
diff --git a/config/external/sqlite/moz.build b/config/external/sqlite/moz.build
index c13061965db1..9c490300b45e 100644
--- a/config/external/sqlite/moz.build
+++ b/config/external/sqlite/moz.build
@@ -2,15 +2,20 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-DIRS += ["../../../third_party/sqlite3/src"]
-if CONFIG["MOZ_FOLD_LIBS"]:
+if CONFIG["MOZ_SYSTEM_SQLITE"]:
     Library("sqlite")
-    # When folding libraries, sqlite is actually in the nss library.
-    USE_LIBS += [
-        "nss",
-    ]
+    DIRS += ["../../../third_party/sqlite3/ext"]
+    OS_LIBS += CONFIG["SQLITE_LIBS"]
 else:
-    SharedLibrary("sqlite")
-    SHARED_LIBRARY_NAME = "mozsqlite3"
+    DIRS += ["../../../third_party/sqlite3/src"]
+    if CONFIG["MOZ_FOLD_LIBS"]:
+        Library("sqlite")
+        # When folding libraries, sqlite is actually in the nss library.
+        USE_LIBS += [
+            "nss",
+        ]
+    else:
+        SharedLibrary("sqlite")
+        SHARED_LIBRARY_NAME = "mozsqlite3"
 
-    SYMBOLS_FILE = "/third_party/sqlite3/src/sqlite.symbols"
+        SYMBOLS_FILE = "/third_party/sqlite3/src/sqlite.symbols"
diff --git a/storage/SQLiteMutex.h b/storage/SQLiteMutex.h
index e2819cc0c0ab..9b76d44697eb 100644
--- a/storage/SQLiteMutex.h
+++ b/storage/SQLiteMutex.h
@@ -54,7 +54,7 @@ class SQLiteMutex : private BlockingResourceBase {
    */
   void lock() {
     MOZ_ASSERT(mMutex, "No mutex associated with this wrapper!");
-#if defined(DEBUG)
+#if defined(DEBUG) && !defined(MOZ_SYSTEM_SQLITE)
     // While SQLite Mutexes may be recursive, in our own code we do not want to
     // treat them as such.
     CheckAcquire();
@@ -62,7 +62,7 @@ class SQLiteMutex : private BlockingResourceBase {
 
     ::sqlite3_mutex_enter(mMutex);
 
-#if defined(DEBUG)
+#if defined(DEBUG) && !defined(MOZ_SYSTEM_SQLITE)
     Acquire();  // Call is protected by us holding the mutex.
 #endif
   }
@@ -72,7 +72,7 @@ class SQLiteMutex : private BlockingResourceBase {
    */
   void unlock() {
     MOZ_ASSERT(mMutex, "No mutex associated with this wrapper!");
-#if defined(DEBUG)
+#if defined(DEBUG) && !defined(MOZ_SYSTEM_SQLITE)
     // While SQLite Mutexes may be recursive, in our own code we do not want to
     // treat them as such.
     Release();  // Call is protected by us holding the mutex.
diff --git a/storage/moz.build b/storage/moz.build
index aa0e82f0d9f7..07646fc0c7ff 100644
--- a/storage/moz.build
+++ b/storage/moz.build
@@ -98,8 +98,12 @@ FINAL_LIBRARY = "xul"
 if CONFIG["MOZ_THUNDERBIRD"] or CONFIG["MOZ_SUITE"]:
     DEFINES["MOZ_SQLITE_FTS3_TOKENIZER"] = 1
 
-# Disable auxiliary files persistence if requested.
-if not CONFIG["MOZ_AVOID_DISK_REMNANT_ON_CLOSE"]:
+# Disable auxiliary files persistence if requested. System SQLite may not have
+# Gecko's compile-time journal size limit, so retain its normal WAL lifecycle.
+if (
+    not CONFIG["MOZ_AVOID_DISK_REMNANT_ON_CLOSE"]
+    and not CONFIG["MOZ_SYSTEM_SQLITE"]
+):
     DEFINES["MOZ_SQLITE_PERSIST_AUXILIARY_FILES"] = 1
 
 # The vector extension is enabled only for Firefox, it could be opened up
@@ -107,12 +111,22 @@ if not CONFIG["MOZ_AVOID_DISK_REMNANT_ON_CLOSE"]:
 if CONFIG["MOZ_BUILD_APP"] == "browser":
     DEFINES["MOZ_SQLITE_VEC0_EXT"] = 1
 
+# The bundled library selects an in-memory temp store at compile time on
+# Android and 64-bit builds. Apply the equivalent per connection when sharing
+# the platform SQLite library.
+if CONFIG["MOZ_SYSTEM_SQLITE"] and (
+    CONFIG["OS_TARGET"] == "Android" or CONFIG["HAVE_64BIT_BUILD"]
+):
+    DEFINES["MOZ_MEMORY_TEMP_STORE_PRAGMA"] = True
+
 LOCAL_INCLUDES += [
     "/dom/base",
     "/third_party/sqlite3/ext",
-    "/third_party/sqlite3/src",
 ]
 
+if not CONFIG["MOZ_SYSTEM_SQLITE"]:
+    LOCAL_INCLUDES += ["/third_party/sqlite3/src"]
+
 if CONFIG["MOZ_FOLD_LIBS"]:
     DEFINES["MOZ_FOLD_LIBS"] = True
 
diff --git a/storage/mozStorageConnection.cpp b/storage/mozStorageConnection.cpp
index a289eee7bc67..062186bf4226 100644
--- a/storage/mozStorageConnection.cpp
+++ b/storage/mozStorageConnection.cpp
@@ -47,6 +47,10 @@
 #include "FileSystemModule.h"
 #include "mozStorageHelper.h"
 
+#if defined(MOZ_SYSTEM_SQLITE) && defined(MOZ_SQLITE_VEC0_EXT)
+#  include "sqlite-vec/sqlite-vec.h"
+#endif
+
 #include "mozilla/Assertions.h"
 #include "mozilla/Logging.h"
 #include "mozilla/Printf.h"
@@ -1416,6 +1420,13 @@ nsresult Connection::initializeInternal() {
     return convertResultCode(srv);
   }
 
+#ifdef MOZ_MEMORY_TEMP_STORE_PRAGMA
+  srv = executeSql(mDBConn, "PRAGMA temp_store = 2;");
+  if (srv != SQLITE_OK) {
+    return convertResultCode(srv);
+  }
+#endif
+
   // Register our built-in SQL functions.
   srv = registerFunctions(mDBConn);
   if (srv != SQLITE_OK) {
@@ -3037,11 +3048,13 @@ Connection::LoadExtension(const nsACString& aExtensionName,
     return NS_ERROR_NOT_INITIALIZED;
   }
 
+#ifndef MOZ_SYSTEM_SQLITE
   int srv = ::sqlite3_db_config(mDBConn, SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION,
                                 1, nullptr);
   if (srv != SQLITE_OK) {
     return NS_ERROR_UNEXPECTED;
   }
+#endif
 
   // Track the loaded extension for later connection cloning operations.
   {
@@ -3056,13 +3069,11 @@ Connection::LoadExtension(const nsACString& aExtensionName,
     }
   }
 
-  nsAutoCString entryPoint("sqlite3_");
-  entryPoint.Append(aExtensionName);
-  entryPoint.AppendLiteral("_init");
+  nsCString extensionName(aExtensionName);
 
   RefPtr<Runnable> loadTask = NS_NewRunnableFunction(
       "mozStorageConnection::LoadExtension",
-      [this, self = RefPtr(this), entryPoint = std::move(entryPoint),
+      [this, self = RefPtr(this), extensionName = std::move(extensionName),
        callback = RefPtr(aCallback)]() mutable {
         MOZ_ASSERT(
             !NS_IsMainThread() ||
@@ -3070,6 +3081,23 @@ Connection::LoadExtension(const nsACString& aExtensionName,
                  eventTargetOpenedOn == GetMainThreadSerialEventTarget()),
             "Should happen on main-thread only for synchronous connections "
             "opened on the main thread");
+#ifdef MOZ_SYSTEM_SQLITE
+        int srv;
+        if (extensionName.EqualsLiteral("fts5")) {
+          // Sailfish's platform SQLite provides FTS5 in the core library.
+          srv = SQLITE_OK;
+#  ifdef MOZ_SQLITE_VEC0_EXT
+        } else if (extensionName.EqualsLiteral("vec")) {
+          srv = ::sqlite3_vec_init(mDBConn, nullptr, nullptr);
+#  endif
+        } else {
+          MOZ_ASSERT_UNREACHABLE("Unsupported SQLite extension");
+          srv = SQLITE_ERROR;
+        }
+#else
+        nsAutoCString entryPoint("sqlite3_");
+        entryPoint.Append(extensionName);
+        entryPoint.AppendLiteral("_init");
 #ifdef MOZ_FOLD_LIBS
         int srv = ::sqlite3_load_extension(mDBConn,
                                            MOZ_DLL_PREFIX "nss3" MOZ_DLL_SUFFIX,
@@ -3078,6 +3106,7 @@ Connection::LoadExtension(const nsACString& aExtensionName,
         int srv = ::sqlite3_load_extension(
             mDBConn, MOZ_DLL_PREFIX "mozsqlite3" MOZ_DLL_SUFFIX,
             entryPoint.get(), nullptr);
+#endif
 #endif
         if (!callback) {
           return;
diff --git a/storage/mozStorageService.cpp b/storage/mozStorageService.cpp
index 6d457dbf2dfc..dfbf13797c25 100644
--- a/storage/mozStorageService.cpp
+++ b/storage/mozStorageService.cpp
@@ -3,6 +3,8 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "BaseVFS.h"
+#include "mozilla/Assertions.h"
+#include "mozilla/Attributes.h"
 #include "mozilla/DebugOnly.h"
 #include "mozilla/SpinEventLoopUntil.h"
 #include "nsIFile.h"
@@ -45,7 +47,7 @@ namespace mozilla::storage {
 ////////////////////////////////////////////////////////////////////////////////
 //// Memory Reporting
 
-#ifdef MOZ_DMD
+#if defined(MOZ_DMD) && !defined(MOZ_SYSTEM_SQLITE)
 mozilla::Atomic<size_t> gSqliteMemoryUsed;
 #endif
 
@@ -142,7 +144,7 @@ Service::CollectReports(nsIHandleReportCallback* aHandleReport,
                  SQLITE_DBSTATUS_SCHEMA_USED, &totalConnSize);
     }
 
-#ifdef MOZ_DMD
+#if defined(MOZ_DMD) && !defined(MOZ_SYSTEM_SQLITE)
     if (::sqlite3_memory_used() != int64_t(gSqliteMemoryUsed)) {
       NS_WARNING(
           "memory consumption reported by SQLite doesn't match "
@@ -320,6 +322,19 @@ nsresult Service::initialize() {
     return convertResultCode(rc);
   }
 
+#ifdef MOZ_SYSTEM_SQLITE
+  constexpr int kMinimumSQLiteVersion = 3053002;
+  MOZ_RELEASE_ASSERT(
+      ::sqlite3_libversion_number() >= kMinimumSQLiteVersion &&
+          ::sqlite3_compileoption_used("THREADSAFE=1") &&
+          ::sqlite3_compileoption_used("SECURE_DELETE") &&
+          ::sqlite3_compileoption_used("ENABLE_UNLOCK_NOTIFY") &&
+          ::sqlite3_compileoption_used("ENABLE_DBSTAT_VTAB") &&
+          ::sqlite3_compileoption_used("ENABLE_CARRAY") &&
+          ::sqlite3_compileoption_used("ENABLE_FTS5"),
+      "System SQLite does not satisfy Gecko storage requirements");
+#endif
+
   /**
    *                    The virtual file system hierarchy
    *
diff --git a/storage/test/unit/test_storage_ext_fts5.js b/storage/test/unit/test_storage_ext_fts5.js
index 7adf4e8b834f..07f54d403dff 100644
--- a/storage/test/unit/test_storage_ext_fts5.js
+++ b/storage/test/unit/test_storage_ext_fts5.js
@@ -14,14 +14,16 @@ add_setup(async function () {
 add_task(async function test_synchronous() {
   info("Testing synchronous connection");
   let conn = getOpenedUnsharedDatabase();
-  Assert.throws(
-    () =>
-      conn.executeSimpleSQL(
-        "CREATE VIRTUAL TABLE email USING fts5(sender, title, body);"
-      ),
-    /NS_ERROR_FAILURE/,
-    "Should not be able to use FTS5 without loading the extension"
-  );
+  if (!AppConstants.MOZ_SYSTEM_SQLITE) {
+    Assert.throws(
+      () =>
+        conn.executeSimpleSQL(
+          "CREATE VIRTUAL TABLE email USING fts5(sender, title, body);"
+        ),
+      /NS_ERROR_FAILURE/,
+      "Should not be able to use FTS5 without loading the extension"
+    );
+  }
 
   await loadFTS5Extension(conn);
 
@@ -59,14 +61,16 @@ add_task(async function test_asynchronous() {
   info("Testing asynchronous connection");
   let conn = await openAsyncDatabase(getTestDB());
 
-  await Assert.rejects(
-    executeSimpleSQLAsync(
-      conn,
-      "CREATE VIRTUAL TABLE email USING fts5(sender, title, body);"
-    ),
-    err => err.message.startsWith("no such module"),
-    "Should not be able to use FTS5 without loading the extension"
-  );
+  if (!AppConstants.MOZ_SYSTEM_SQLITE) {
+    await Assert.rejects(
+      executeSimpleSQLAsync(
+        conn,
+        "CREATE VIRTUAL TABLE email USING fts5(sender, title, body);"
+      ),
+      err => err.message.startsWith("no such module"),
+      "Should not be able to use FTS5 without loading the extension"
+    );
+  }
 
   await loadFTS5Extension(conn);
 
diff --git a/third_party/sqlite3/ext/moz.build b/third_party/sqlite3/ext/moz.build
index ee80c3a0811a..82a6df7da088 100644
--- a/third_party/sqlite3/ext/moz.build
+++ b/third_party/sqlite3/ext/moz.build
@@ -9,17 +9,26 @@ NoVisibilityFlags()
 # We allow warnings for third-party code that can be updated from upstream.
 AllowCompilerWarnings()
 
-if CONFIG["MOZ_FOLD_LIBS"]:
+if CONFIG["MOZ_SYSTEM_SQLITE"]:
+    FINAL_LIBRARY = "sqlite"
+elif CONFIG["MOZ_FOLD_LIBS"]:
     # When folding libraries, sqlite is actually in the nss library.
     FINAL_LIBRARY = "nss"
 else:
     # The final library is in config/external/sqlite
     FINAL_LIBRARY = "sqlite"
 
-SOURCES += [
-    "fts5.c",
-    "sqlite-vec/sqlite-vec.c",
-]
+if CONFIG["MOZ_SYSTEM_SQLITE"]:
+    SOURCES += [
+        "sqlite-vec/sqlite-vec.c",
+    ]
+    DEFINES["SQLITE_CORE"] = 1
+    CFLAGS += CONFIG["SQLITE_CFLAGS"]
+else:
+    SOURCES += [
+        "fts5.c",
+        "sqlite-vec/sqlite-vec.c",
+    ]
 
 # gcc had issues with current Neon code:
 # https://bugzilla.mozilla.org/show_bug.cgi?id=2041214
diff --git a/toolkit/library/moz.build b/toolkit/library/moz.build
index 0222cc487730..57af0569c4bc 100644
--- a/toolkit/library/moz.build
+++ b/toolkit/library/moz.build
@@ -160,6 +160,9 @@ USE_LIBS += [
     "sqlite",
 ]
 
+if CONFIG["MOZ_SYSTEM_SQLITE"]:
+    OS_LIBS += CONFIG["SQLITE_LIBS"]
+
 if not CONFIG["USE_LIBZ_RS"]:
     USE_LIBS += [
         "zlib",
diff --git a/toolkit/modules/AppConstants.sys.mjs b/toolkit/modules/AppConstants.sys.mjs
index b9d2d8e0d707..2e028a57ef1d 100644
--- a/toolkit/modules/AppConstants.sys.mjs
+++ b/toolkit/modules/AppConstants.sys.mjs
@@ -145,6 +145,13 @@ export var AppConstants = Object.freeze({
 
   MOZ_SYSTEM_NSS: @MOZ_SYSTEM_NSS_BOOL@,
 
+  MOZ_SYSTEM_SQLITE:
+#ifdef MOZ_SYSTEM_SQLITE
+    true,
+#else
+    false,
+#endif
+
   MOZ_PLACES: @MOZ_PLACES_BOOL@,
 
   MOZ_REQUIRE_SIGNING: @MOZ_REQUIRE_SIGNING_BOOL@,
@@ -261,7 +268,9 @@ export var AppConstants = Object.freeze({
   MOZ_SELECTABLE_PROFILES: @MOZ_SELECTABLE_PROFILES_BOOL@,
 
   SQLITE_LIBRARY_FILENAME:
-#ifdef MOZ_FOLD_LIBS
+#ifdef MOZ_SYSTEM_SQLITE
+  "",
+#elif MOZ_FOLD_LIBS
   "@DLL_PREFIX@nss3@DLL_SUFFIX@",
 #else
   "@DLL_PREFIX@mozsqlite3@DLL_SUFFIX@",
diff --git a/toolkit/moz.configure b/toolkit/moz.configure
index fe8668f98383..65eda6f5804f 100644
--- a/toolkit/moz.configure
+++ b/toolkit/moz.configure
@@ -3223,6 +3223,10 @@ with only_when(compile_environment & target_has_linux_kernel):
 # ==============================================================
 include("../build/moz.configure/nss.configure")
 
+# SQLite
+# ==============================================================
+include("../build/moz.configure/sqlite.configure")
+
 
 # Enable or disable running in background task mode: headless for
 # periodic, short-lived, maintenance tasks.
diff --git a/toolkit/xre/AutoSQLiteLifetime.cpp b/toolkit/xre/AutoSQLiteLifetime.cpp
index 83cb76cdb88a..cf20e451e07a 100644
--- a/toolkit/xre/AutoSQLiteLifetime.cpp
+++ b/toolkit/xre/AutoSQLiteLifetime.cpp
@@ -7,7 +7,7 @@
 #include "sqlite3.h"
 #include "mozilla/Atomics.h"
 
-#ifdef MOZ_MEMORY
+#if defined(MOZ_MEMORY) && !defined(MOZ_SYSTEM_SQLITE)
 #  include "mozmemory.h"
 #  ifdef MOZ_DMD
 #    include "nsIMemoryReporter.h"
@@ -115,7 +115,7 @@ const sqlite3_mem_methods memMethods = {
 
 }  // namespace
 
-#endif  // MOZ_MEMORY
+#endif  // MOZ_MEMORY && !MOZ_SYSTEM_SQLITE
 
 namespace mozilla {
 
@@ -129,11 +129,16 @@ AutoSQLiteLifetime::AutoSQLiteLifetime() {
 
 // static
 void AutoSQLiteLifetime::Init() {
-#ifdef MOZ_MEMORY
-  sResult = ::sqlite3_config(SQLITE_CONFIG_MALLOC, &memMethods);
+#ifdef MOZ_SYSTEM_SQLITE
+  // SQLite is shared with the embedding process. It may already be initialized
+  // by Qt, so Gecko must not replace its allocator or take global ownership.
+  sResult = SQLITE_OK;
 #else
+#  ifdef MOZ_MEMORY
+  sResult = ::sqlite3_config(SQLITE_CONFIG_MALLOC, &memMethods);
+#  else
   sResult = SQLITE_OK;
-#endif
+#  endif
 
   if (sResult == SQLITE_OK) {
     // TODO (bug 1191405): do not preallocate the connections caches until we
@@ -145,14 +150,17 @@ void AutoSQLiteLifetime::Init() {
     // the documentation suggests calling this directly.  So we do.
     sResult = ::sqlite3_initialize();
   }
+#endif
 }
 
 AutoSQLiteLifetime::~AutoSQLiteLifetime() {
+#ifndef MOZ_SYSTEM_SQLITE
   // Shutdown the sqlite3 API.  Warn if shutdown did not turn out okay, but
   // there is nothing actionable we can do in that case.
   sResult = ::sqlite3_shutdown();
   NS_WARNING_ASSERTION(sResult == SQLITE_OK,
                        "sqlite3 did not shutdown cleanly.");
+#endif
 }
 
 int AutoSQLiteLifetime::sSingletonEnforcer = 0;
diff --git a/toolkit/xre/moz.build b/toolkit/xre/moz.build
index 114fad146337..3065155e976e 100644
--- a/toolkit/xre/moz.build
+++ b/toolkit/xre/moz.build
@@ -270,6 +270,8 @@ LOCAL_INCLUDES += [
 if CONFIG["MOZ_EMBEDLITE"]:
     LOCAL_INCLUDES += ["/mobile/sailfishos/embedprocess"]
 
+CXXFLAGS += CONFIG["SQLITE_CFLAGS"]
+
 if CONFIG["MOZ_SANDBOX"] and CONFIG["OS_ARCH"] == "WINNT":
     LOCAL_INCLUDES += [
         "/security/sandbox/chromium",
diff --git a/tools/@types/subs/AppConstants.sys.d.mts b/tools/@types/subs/AppConstants.sys.d.mts
index 0d4c07999734..6e62ddc74be0 100644
--- a/tools/@types/subs/AppConstants.sys.d.mts
+++ b/tools/@types/subs/AppConstants.sys.d.mts
@@ -88,6 +88,8 @@ export const AppConstants: Readonly<{
 
   MOZ_SYSTEM_NSS: boolean;
 
+  MOZ_SYSTEM_SQLITE: boolean;
+
   MOZ_PLACES: boolean;
 
   MOZ_REQUIRE_SIGNING: boolean;
